sharding/shard.go

```diff
@@ -0,0 +1,262 @@
+package sharding
+
+import (
+	"bytes"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// ShardBackend defines an interface for a shardDB's necessary method
+// signatures.
+type ShardBackend interface {
+	Get(k common.Hash) (*[]byte, error)
+	Has(k common.Hash) bool
+	Put(k common.Hash, val []byte) error
+	Delete(k common.Hash) error
+}
+
+// Shard base struct.
+type Shard struct {
+	shardDB ShardBackend
+	shardID *big.Int
+}
+
+// NewShard creates an instance of a Shard struct given a shardID.
+func NewShard(shardID *big.Int, shardDB ShardBackend) *Shard {
+	return &Shard{
+		shardID: shardID,
+		shardDB: shardDB,
+	}
+}
+
+// ShardID gets the shard's identifier.
+func (s *Shard) ShardID() *big.Int {
+	return s.shardID
+}
+
+// ValidateShardID checks if header belongs to shard.
+func (s *Shard) ValidateShardID(h *CollationHeader) error {
+	if s.ShardID().Cmp(h.ShardID()) != 0 {
+		return fmt.Errorf("collation does not belong to shard %d but instead has shardID %d", h.ShardID().Int64(), s.ShardID().Int64())
+	}
+	return nil
+}
+
+// HeaderByHash looks up a collation header from the shardDB using the header's hash.
+func (s *Shard) HeaderByHash(hash *common.Hash) (*CollationHeader, error) {
+	encoded, err := s.shardDB.Get(*hash)
+	if err != nil {
+		return nil, fmt.Errorf("get failed: %v", err)
+	}
+	if encoded == nil {
+		return nil, fmt.Errorf("no value set for header hash: %v", hash.String())
+	}
+
+	var header CollationHeader
+
+	stream := rlp.NewStream(bytes.NewReader(*encoded), uint64(len(*encoded)))
+	if err := header.DecodeRLP(stream); err != nil {
+		return nil, fmt.Errorf("could not decode RLP header: %v", err)
+	}
+
+	return &header, nil
+}
+
+// CollationByHash fetches full collation.
+func (s *Shard) CollationByHash(headerHash *common.Hash) (*Collation, error) {
+	header, err := s.HeaderByHash(headerHash)
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch header by hash: %v", err)
+	}
+
+	body, err := s.BodyByChunkRoot(header.ChunkRoot())
+	if err != nil {
+		return nil, fmt.Errorf("cannot fetch body by chunk root: %v", err)
+	}
+	// TODO: deserializes the body into a txs slice instead of using
```

Comment:
Please create github issue for this. It would be a nice place for someone to contribute after serialization is added.

Reply:
I'll create as soon as we merge